### ISSUE TITLE
`=Y`, `~Y` should match individual keywords, not the space-separated list of keywords

### ISSUE THREAD

**Issue by falsifian** [neomutt/neomutt#4360]
I've been diving into the neomutt source code recently so may try writing a patch if it's likely to be accepted.

## What would you like to be able to do?

I'd like to be able to limit to messages with the "s" IMAP keyword using the limit pattern `~Y^s$`.

Unfortunately that doesn't work right now because the pattern is matched against the whole string. So if my email has keywords "r" and "s" the pattern gets matched against the string "s r". (Presumably "r s" is possible too, but that's not what neomutt ended up with in my case.)

I suggest the meaning of `=Y` and `~Y` change to: at least one of the keywords matches the pattern.

This might annoy people who were relying on the old behaviour. But the old behaviour seems undefined since the order in which the keywords appear in the string is unspecified.

## What needs to change in NeoMutt?

I am guessing (hoping) this is a simple change to the `MUTT_PAT_DRIVER_TAGS` case in `pattern_exec` in `pattern/exec.c`, to break the tag list into words and try each separately. However I don't understand why `MUTT_PAT_DRIVER_TAGS` appears in the `mutt_pattern_alias_exec` function (I don't really understand what that function does) so there's a piece of the picture I'm missing.

## Have you tried any alternative methods?

I can instead limit to `~Y"(^| )s($| )"` but that's annoying.

**Comment by falsifian** [comment]
Actually, I'd also like it if `=Yfoo` matched only messages with the keyword `foo`. Currently it will also match messages with `foobar`. At least in my use I have never wanted to match keywords by anything other than the entire keyword. Substrings could still be matched with `~Yfoo`.

I guess that would make `=Y` inconsistent with `=B`, `=b`, `=C` etc which all match substrings, but there's already precedent for pattern modifiers to behave differently: `=h`.

**Comment by gahr** [comment]
Hi @falsifian , I'll try to take a look

**Comment by gahr** [comment]
Can you give the branch devel/issue-4360 a go?

**Comment by gahr** [comment]
As per `%Y foo` matching `foobar`, you can use `~Y "^foo$"`

**Comment by falsifian** [comment]
Thanks @gahr.

With that version, `!~Y^r$` matches a message that has two tags, `r` and `s`. It would make more sense to me (and for my use case) if `!Y^r$` meant "messages that don't have the tag `r`.

BTW, do you have any idea why `MUTT_PAT_DRIVER_TAGS` appears in `mutt_pattern_alias_exec`? Should that be changed too? I don't know what it's for.

Sure, it makes sense to make `=Y` meaning `~Y^foo$` a separate issue. Let's get this one done first.

**Comment by gahr** [comment]
> With that version, `!~Y^r$` matches a message that has two tags, `r` and `s`

Good point, I'll get back to work :)

**Comment by gahr** [comment]
Fixed in de7c0387c76383b1917d8fed2f3c1bec365df216

**Comment by gahr** [comment]
> BTW, do you have any idea why `MUTT_PAT_DRIVER_TAGS` appears in `mutt_pattern_alias_exec`? Should that be changed too? I don't know what it's for.

It is very poorly (i.e., not) documented, but aliases can have tags, and you can use `~Y` to query them.
```
:alias myself me@example.com # some comment tags:foo,bar,.baz
```

So, yeah, I think I should fix them the same way
